# Credit Risk using Naive bayes and KNN

# Author- Rohan Kulkarni

In [1]:
import pandas as pd

df = pd.read_csv(r"C:/Users/_RoXstaR_/OneDrive/Desktop/CreditRisk.csv")
df.Credit_History = df.Credit_History.fillna(1)
df.Gender = df.Gender.fillna("Male")
df.Married = df.Married.fillna("No")
df.Dependents = df.Dependents.fillna(0)
df.Self_Employed = df.Self_Employed.fillna("Yes")
df.LoanAmount = df.LoanAmount.fillna( df.LoanAmount.mean() )
df.Loan_Amount_Term = df.Loan_Amount_Term.fillna(df.Loan_Amount_Term.mean())




In [2]:

from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()

In [3]:
df.Loan_Status = le.fit_transform(df.Loan_Status)

df.Gender = le.fit_transform(df.Gender)
df.Married = le.fit_transform(df.Married)
df.Education = le.fit_transform(df.Education)
df.Self_Employed = le.fit_transform(df.Self_Employed)
df.Property_Area = le.fit_transform(df.Property_Area)

In [4]:
df1 = df
from sklearn.model_selection import train_test_split
df = df.iloc[: , 1::]
df_train , df_test = train_test_split(df, test_size = .2)

In [5]:
df_train_x = df_train.iloc[:,0:-1]
df_train_y=df_train.iloc[:,-1]

In [6]:
df_test_x = df_test.iloc[:,0:-1]
df_test_y=df_test.iloc[:,-1]

## Naive bayes

In [7]:
from sklearn.naive_bayes import MultinomialNB

In [8]:
nb_cr=MultinomialNB()

In [9]:
nb_cr.fit(df_train_x,df_train_y)

MultinomialNB()

In [10]:
pred_nb_cr=nb_cr.predict(df_test_x)

In [11]:
from sklearn.metrics import confusion_matrix

In [12]:
confusion_matrix(pred_nb_cr,df_test_y)

array([[28, 72],
       [27, 70]], dtype=int64)

## Decision Tree

In [13]:
from sklearn.tree import DecisionTreeClassifier

In [14]:
dt=DecisionTreeClassifier()

In [15]:
dt.fit(df_train_x,df_train_y)

DecisionTreeClassifier()

In [16]:
pred_dt=dt.predict(df_test_x)

In [17]:
from sklearn.metrics import confusion_matrix

In [18]:
confusion_matrix(pred_dt,df_test_y)

array([[ 28,  21],
       [ 27, 121]], dtype=int64)

## Random Forest

In [19]:
from sklearn.ensemble import RandomForestClassifier

In [20]:
rf=RandomForestClassifier()

In [21]:
rf.fit(df_train_x,df_train_y)

RandomForestClassifier()

In [22]:
pred_rf=rf.predict(df_test_x)

In [23]:
confusion_matrix(pred_rf,df_test_y)

array([[ 25,   5],
       [ 30, 137]], dtype=int64)

## Logistic Regression

In [24]:
from sklearn.linear_model import LogisticRegression

In [25]:
lr=LogisticRegression()

In [26]:
lr.fit(df_train_x,df_train_y)

C:\Users\_RoXstaR_\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:762: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [27]:
pred_lr=lr.predict(df_test_x)

In [28]:
confusion_matrix(pred_lr,df_test_y)

array([[ 20,   1],
       [ 35, 141]], dtype=int64)

## KNN

In [29]:
from sklearn.neighbors import KNeighborsClassifier

In [30]:
knn=KNeighborsClassifier(n_neighbors=7)

In [31]:
knn.fit(df_train_x,df_train_y)

KNeighborsClassifier(n_neighbors=7)

In [32]:
pred_knn=knn.predict(df_test_x)

In [33]:
confusion_matrix(pred_knn,df_test_y)

array([[  9,   8],
       [ 46, 134]], dtype=int64)

In [34]:
from sklearn.metrics import accuracy_score

In [35]:
print('Accuracy of Naive Bayes :',accuracy_score(df_test_y,pred_nb_cr))
print('-------------------------')

print('Accuracy of Decision Tree :',accuracy_score(df_test_y,pred_dt))
print('-------------------------')

print('Accuracy of Random Forest :',accuracy_score(df_test_y,pred_rf))
print('-------------------------')

print('Accuracy of Logistic Regression :',accuracy_score(df_test_y,pred_lr))
print('-------------------------')

print('Accuracy of KNN :',accuracy_score(df_test_y,pred_knn))
print('-------------------------')

Accuracy of Naive Bayes : 0.49746192893401014
-------------------------
Accuracy of Decision Tree : 0.7563451776649747
-------------------------
Accuracy of Random Forest : 0.8223350253807107
-------------------------
Accuracy of Logistic Regression : 0.817258883248731
-------------------------
Accuracy of KNN : 0.7258883248730964
-------------------------


In [ ]:
from sklearn.svm import SVC

In [68]:
svc_cr=SVC(kernel='linear')

In [69]:
svc_cr.fit(df_train_x,df_train_y)

SVC(kernel='linear')

In [70]:
pred_svc=svc_cr.predict(df_test_x)

In [71]:
confusion_matrix(pred_svc,df_test_y)

array([[ 20,   2],
       [ 35, 140]], dtype=int64)

1. with sigmoid kernel
       [  7,  28],
       [ 48, 114]],
       
2. with polynomial kernel
       [  0,   0],
       [ 55, 142]]
       
3. with rbf kernel
        [  0,   0],
       [ 55, 142]]
       
4. with linear kernel
        [[ 20,   2],
        [ 35, 140]

##### looks like Linear is performing well than any other kernel, lets do our grid search

##### GRID

In [72]:
search_dict={'kernel':('linear','poly','rbf','sigmoid'),}

In [74]:
from sklearn.model_selection import GridSearchCV

In [77]:
grid=GridSearchCV(svc_cr,param_grid=search_dict)

In [ ]:
grid.fit(df_train_x,df_train_y)